### ISSUE TITLE
Allow checking various metadata by fixtures in end2end tests

### ISSUE THREAD

**Issue by ctiller** [grpc/grpc#59]
1. have the fixture supply the client/server cq_verifier
2. add something to the cq verifier to add metadata checks to particular events when they actually come in

We want to be able to configure a cq verifier to be able to say:
'whenever I get one of client_metadata_read, server_rpc_new, or finished, I expect these pieces of metadata to be present'

And we want the fixture to be able to configure this - for verifying oauth code.

**Comment by ctiller** [comment]
Removing from GA: I'm not sure we need this for the coverage story.

**Comment by mehrdada** [comment]
Closing as stale. Please re-open if of concern today.

**Comment by ctiller** [comment]
This is still needed (moreso today than ever) and will probably get addressable when we start transitioning to C++.

**Comment by ctiller** [comment]
Nowadays this could be rephrased as:

1. write a helper library to instantiate filters by themselves (we do this in some benchmarks)
2. write mock libraries so that we can declare useful expectations
3. write injection libraries to push ops into filters and finish them below

These three pieces together should get us a long way to being able to test filters in isolation

WDYT @markdroth

**Comment by ctiller** [comment]
Reassigning in case this is something @AspirinSJL thinks is interesting

**Comment by markdroth** [comment]
Yes, I think this would be a great idea.  It sounds like this would actually be a framework for unit testing individual filters, which sounds really useful -- it would give us a way to directly test things like callbacks happening in unexpected orders.

I've updated the description of this bug accordingly.

I agree that it should wait until after we transition core to C++, since that will allow us to make use of gmock.

**Comment by AspirinSJL** [comment]
First question: what does transitioning core to C++ mean?

**Comment by markdroth** [comment]
https://github.com/grpc/proposal/blob/master/L6-allow-c%2B%2B-in-grpc-core.md

**Comment by TheNumberDevil** [comment]
Guys is this issue still valid, Can I look in to it

**Comment by markdroth** [comment]
This is still valid, but it probably doesn't make sense to tackle this now, since I want to make some significant changes to the filter API before we do this.

**Comment by ctiller** [comment]
Assuming you won't object given current interests Mark

**Comment by markdroth** [comment]
No objection.  Thanks!

**Comment by markdroth** [comment]
@pawbhard, I think you are already working on this.
